dashboard/app/models/sections/section_instructor.rb

```diff
@@ -33,3 +33,33 @@ class SectionInstructor < ApplicationRecord
     removed: 3,
   }
 end
+
+public def create_section(section, email)
+  # Enforce maximum co-instructor count (the limit is 5 plus the main teacher
+  # for a total of 6)
+  if SectionInstructor.where(section: section).count >= 6
+    raise ArgumentError.new('section full')
+  end
+
+  instructor = User.find_by(email: email, user_type: :teacher)
+
+  si = SectionInstructor.with_deleted.find_by(instructor: instructor, section: section)
+  if instructor.blank?
+    raise
+  end
+
+  # Actually delete the instructor if they were soft-deleted so they can be re-invited.
+  if si&.deleted_at.present?
+    si.really_destroy!
+  # Can't re-add someone who is already an instructor (or invited/declined)
+  elsif si.present?
+    raise ArgumentError.new('already invited')
+  end
+
+  return SectionInstructor.create!(
+    section: section,
+    instructor: instructor,
+    status: :invited,
+    invited_by: current_user
+  )
+end
```

Comment:
This method feels like it fits better in `section.rb`. You're adding an instructor to a section, so the method can/should be owned by the section. Moving it over there, it would look something like below.

A couple of changes to point out:
1) This now uses your new `section.section_instructors` attribute instead of querying the `SectionInstructor` model directly.
2) `find_by!` raises an exception if a record can't be found so you don't need to manually raise if blank,
3) In Ruby the last line of a method is automatically returned, so the `return` keyword is optional and conventionally omitted in this case.

```suggestion
public def add_instructor(email)
  # Enforce maximum co-instructor count (the limit is 5 plus the main teacher
  # for a total of 6)
  if section_instructors.count >= 6
    raise ArgumentError.new('section full')
  end

  instructor = User.find_by!(email: email, user_type: :teacher)

  si = SectionInstructor.with_deleted.find_by(instructor: instructor, section: section)

  # Actually delete the instructor if they were soft-deleted so they can be re-invited.
  if si&.deleted?
    si.really_destroy!
  # Can't re-add someone who is already an instructor (or invited/declined)
  elsif si.present?
    raise ArgumentError.new('already invited')
  end

 SectionInstructor.create!(
    section: self,
    instructor: instructor,
    status: :invited,
    invited_by: current_user
  )
end
```